acceptance/tests/options/config_file/ttls_cached_custom_and_core_fact.rb

```diff
@@ -0,0 +1,141 @@
+test_name 'missing facts should not invalidate cache' do
+  tag 'risk:high'
+
+  require 'facter/acceptance/user_fact_utils'
+  extend Facter::Acceptance::UserFactUtils
+
+  custom_fact_file = 'custom_fact.rb'
+  custom_fact_name = 'my_custom_fact'
+  custom_cache_group = 'my_custom_group'
+  custom_fact_value = 'banana'
+  core_fact_name = 'system_uptime.seconds'
+
+  fact_content = <<-CUSTOM_FACT
+    Facter.add('#{custom_fact_name}') do
+      setcode do
+        "#{custom_fact_value}"
+      end
+    end
+  CUSTOM_FACT
+
+
+  agents.each do |agent|
+    cache_folder = get_cached_facts_dir(agent['platform'], on(agent, facter('kernelmajversion')).stdout.chomp.to_f)
+    fact_dir = agent.tmpdir('facter')
+    env = { 'FACTERLIB' => fact_dir }
+
+    config_dir = get_default_fact_dir(agent['platform'], on(agent, facter('kernelmajversion')).stdout.chomp.to_f)
+    config_file = File.join(config_dir, 'facter.conf')
+
+    config_data = <<~FACTER_CONF
+      facts : {
+        ttls : [
+          { "#{custom_cache_group}" : 3 days }
+        ]
+      }
+
+      fact-groups : {
+        "#{custom_cache_group}" : ["#{custom_fact_name}", "#{core_fact_name}"],
+      }
+    FACTER_CONF
+
+    step "Agent #{agent}: create config file" do
+      agent.mkdir_p(config_dir)
+      create_remote_file(agent, config_file, config_data)
+
+      fact_file = File.join(fact_dir, custom_fact_file)
+      create_remote_file(agent, fact_file, fact_content)
+    end
+
+    teardown do
+      agent.rm_rf(fact_dir)
+      agent.rm_rf("#{cache_folder}/*")
+      agent.rm_rf(config_file)
+    end
+
+    step 'request the core fact' do
+      core_value = on(agent, facter(core_fact_name, environment: env)).stdout.strip.to_i
+
+      cat_output = agent.cat("#{cache_folder}/#{custom_cache_group}")
+      cache = JSON.parse(cat_output)
+
+      step 'check if it is the only fact on cache' do
+        cached = { core_fact_name => core_value, 'cache_format_version' => 1 }
+        assert_equal(cached, cache)
+      end
+
+      step 'check that it cached the value it printed' do
+        assert_equal(core_value, cache[core_fact_name])
+      end
+    end
+
+    step 'request the core fact again' do
+      core_value = on(agent, facter(core_fact_name, environment: env)).stdout.strip.to_i
+
+      cat_output = agent.cat("#{cache_folder}/#{custom_cache_group}")
+      cache = JSON.parse(cat_output)
+
+      step 'check that it cached the value it printed' do
+        assert_equal(core_value, cache[core_fact_name])
```

Comment:
Here you should check that the cached value coincides with the one retrieved in the first request. Making the check like this, you can't be sure that facter doesn't simply solve the core fact again and rewrites the cache file.